File: lintcode/python/1467_ask_for_cooling_time.py
#!/usr/bin/python -t

# hash table
# 数组记录每个技能之前一次的释放时间，然后对于当前技能，计算cd。

class Solution:
    """
    @param arr: The release order
    @param n: The cooldown
    @return: Return the time
    """
    def askForCoolingTime(self, arr, n):
        # Write your code here
        m = len(arr)
        d = {}
        ret = 0
        
        for i in range(m):
            if (arr[i] not in d) or (ret - d[arr[i]] >= n+1):
                ret += 1
            else:
                ret = d[arr[i]] + n + 1
            d[arr[i]] = ret
                
        return ret
        

File: lintcode/golang/1467_ask_for_cooling_time.go
/**
 * @param arr: The release order
 * @param n: The cooldown
 * @return: Return the time
 */
func askForCoolingTime (arr []int, n int) int {
    // Write your code here
    l := len(arr)
    if l == 0 {
        return 0
    }

    d := make(map[int]int)
    ret := 0

    for _, v := range arr {
        ret++
        if pre, ok := d[v]; ok {
            if ret - pre <= n {
                ret = pre+n+1
            }
        }
        d[v] = ret
    }

    return ret
}

